### PR TITLE
[Fix] Refactor axis validation and canonicalization for targeted review

### PR DESCRIPTION
**Pull request by ChiragSW**
Fixes part 2 of #22595 

## Contributor Agreement
**Please check all boxes below before submitting your PR for review:**

- [x] I am a human, and not a bot.
- [x] I will be responsible for responding to review comments in a timely manner.
- [x] I will work with the maintainers to push this PR forward until submission.

> **Note:** Failing to adhere to this agreement may result in your future PRs no longer being reviewed.

### CODE REVIEW

**Comment by ChiragSW** [comment]
@hertschuh please review

File: keras/src/backend/common/backend_utils.py [hunk 1]
```diff
@@ -302,6 +302,35 @@ def canonicalize_axis(axis, num_dims):
...
+            "Argument `axis` must be an integer or a sequence of integers. "
+            f"Received: axis={axis}"
+        )
+
+    for a in axis:
+        if not isinstance(a, int):
+            raise TypeError(
+                "Argument `axis` must be an integer or a sequence of "
+                f"integers. Received: axis={axis}"
+            )
```
Comment (hertschuh) [inline keras/src/backend/common/backend_utils.py]: This part is done in `canonicalize_axis` already, it's `operator.index(axis)`, but maybe it could raise a cleaner error message.

File: keras/src/ops/nn.py [hunk 2]
```diff
@@ -17,6 +19,40 @@
...
+            if not isinstance(a, int):
+                raise TypeError(
+                    "Argument `axis` must be an integer or a sequence of "
+                    f"integers. Received: axis={axis}"
+                )
+        return tuple(axis)
+    raise TypeError(
+        "Argument `axis` must be an integer or a sequence of integers. "
+        f"Received: axis={axis}"
+    )
```
Comment (hertschuh) [inline keras/src/ops/nn.py]: We don't support tensors with a dynamic/unknown rank, so remove and just use `canonicalize_axes`

File: keras/src/ops/operation_utils.py [hunk 3]
```diff
@@ -393,6 +393,24 @@ def reduce_shape(shape, axis=None, keepdims=False):
...
+    shape = getattr(x, "shape", None)
+    if shape is None:
+        return None
+    rank = getattr(shape, "rank", None)
+    if rank is not None:
+        return rank
+    try:
+        return len(shape)
+    except TypeError:
+        return None
```
Comment (hertschuh) [inline keras/src/ops/operation_utils.py]: Remove, we don't support dynamic/unknown ranks. Just use `len(x.shape)` everywhere.

File: keras/src/ops/nn.py [hunk 4]
```diff
@@ -17,6 +19,40 @@
...
+
+def _normalize_axis_for_loss(output, axis):
+    if not isinstance(axis, int):
+        raise TypeError(
+            f"Argument `axis` must be an integer. Received: axis={axis}"
+        )
+    ndim = operation_utils.get_static_tensor_ndim(output)
+    if ndim is not None:
+        return canonicalize_axis(axis, ndim)
+    return axis
```
Comment (hertschuh) [inline keras/src/ops/nn.py]: Same here, remove and just use `canonicalize_axis`.

File: keras/src/ops/nn.py [hunk 5]
```diff
@@ -1014,9 +1050,22 @@ def log_softmax(x, axis=-1):
...
     """
+    axis = _normalize_log_softmax_axis(x, axis)
     if any_symbolic_tensors((x,)):
         return LogSoftmax(axis).symbolic_call(x)
+    x = backend.convert_to_tensor(x)
+    original_shape = x.shape
+    if axis is None:
+        x = backend.numpy.reshape(x, (-1,))
+        x = backend.nn.log_softmax(x, axis=-1)
+        return backend.numpy.reshape(x, original_shape)
```
Comment (hertschuh) [inline keras/src/ops/nn.py]: There should not be any code here, the backend op should handle this.

Which backend has this issue?

File: keras/src/ops/nn.py
Comment (hertschuh) [inline keras/src/ops/nn.py]: Hmm... I'm not sure why this was done this way, but it seems wrong. For one thing, the shape of `x` may not be known at this stage.

This needs to move in each backend specific implementation. I suspect most backends won't need it.

File: keras/src/ops/nn.py [hunk 6]
```diff
@@ -1074,8 +1123,19 @@ def sparsemax(x, axis=-1):
     array([0., 0., 1.], shape=(3,), dtype=float64)
 
     """
+    if not isinstance(axis, int):
+        raise TypeError(
+            f"Argument `axis` must be an integer. Received: axis={axis}"
+        )
+    ndim = operation_utils.get_static_tensor_ndim(x)
+    if ndim is not None:
+        axis = canonicalize_axis(axis, ndim)
```
Comment (hertschuh) [inline keras/src/ops/nn.py]: Remove, this cannot happen here. It should be in the backend specific implementation if it's not already handled.

File: keras/src/ops/nn.py [hunk 7]
```diff
@@ -1074,8 +1123,19 @@ def sparsemax(x, axis=-1):
...
     if any_symbolic_tensors((x,)):
         return Sparsemax(axis).symbolic_call(x)
+    x = backend.convert_to_tensor(x)
+    ndim = operation_utils.get_static_tensor_ndim(x)
+    if ndim is not None:
+        axis = canonicalize_axis(axis, ndim)
```
Comment (hertschuh) [inline keras/src/ops/nn.py]: Remove, this cannot happen here. It should be in the backend specific implementation if it's not already handled.

File: keras/src/ops/nn.py [hunk 8]
```diff
@@ -2183,9 +2248,19 @@ class labels instead of one-hot encoded vectors. It measures the
...
     """
     if any_symbolic_tensors((target, output)):
+        if not isinstance(axis, int):
+            raise TypeError(
+                f"Argument `axis` must be an integer. Received: axis={axis}"
+            )
+        if axis != -1:
+            raise ValueError(
+                f"Only axis=-1 is currently supported. Received: axis={axis}"
+            )
```
Comment (hertschuh) [inline keras/src/ops/nn.py]: Remove, you fixed `compute_output_spec`.

File: keras/src/ops/nn.py [hunk 9]
```diff
@@ -2183,9 +2248,19 @@ class labels instead of one-hot encoded vectors. It measures the
...
         return SparseCategoricalCrossentropy(
             from_logits=from_logits, axis=axis
         ).symbolic_call(target, output)
+    output = backend.convert_to_tensor(output)
+    axis = _normalize_axis_for_loss(output, axis)
```
Comment (hertschuh) [inline keras/src/ops/nn.py]: Remove, this shouldn't be here. If needed add a `canonicalize_axis` in some backend implementations.

File: keras/src/ops/nn_test.py [hunk 10]
```diff
@@ -212,6 +212,22 @@ def test_sparsemax(self):
         x = KerasTensor([None, 2, 3])
         self.assertEqual(knn.sparsemax(x).shape, (None, 2, 3))
 
+    def test_log_softmax_invalid_axis_keras_input(self):
+        x = keras.Input(shape=(3,))
```
Comment (hertschuh) [inline keras/src/ops/nn_test.py]: Use `KerasTensor` instead of `keras.Input` everywhere.

File: keras/src/ops/numpy.py [hunk 11]
```diff
@@ -302,7 +315,7 @@ def all(x, axis=None, keepdims=False):
 
 
 class AllClose(Operation):
-    def __init__(self, rtol=1e-5, atol=1e-8, equal_nan=False, *, name=None):
+    def __init__(self, rtol=1e-05, atol=1e-08, equal_nan=False, *, name=None):
```
Comment (hertschuh) [inline keras/src/ops/numpy.py]: Undo.

File: keras/src/ops/numpy.py [hunk 12]
```diff
@@ -322,7 +335,7 @@ def compute_output_spec(self, x1, x2):
 
 
 @keras_export(["keras.ops.allclose", "keras.ops.numpy.allclose"])
-def allclose(x1, x2, rtol=1e-5, atol=1e-8, equal_nan=False):
+def allclose(x1, x2, rtol=1e-05, atol=1e-08, equal_nan=False):
```
Comment (hertschuh) [inline keras/src/ops/numpy.py]: Undo.

File: keras/src/ops/numpy.py [hunk 13]
```diff
@@ -2682,16 +2700,29 @@ def call(self, x):
         )
 
     def compute_output_spec(self, x):
+        ndim = get_static_tensor_ndim(x)
+        if ndim is None:
+            raise ValueError(
+                "`diagonal` requires a known input rank for symbolic shape "
+                "inference. Received: x.shape=None"
+            )
```
Comment (hertschuh) [inline keras/src/ops/numpy.py]: Remove, KerasTensors always have a rank

File: keras/src/ops/numpy.py [hunk 14]
```diff
@@ -2682,16 +2700,29 @@ def call(self, x):
...
             raise ValueError(
                 "`diagonal` requires an array of at least two dimensions, but "
                 f"`x` is of shape {x.shape}."
             )
+        axis1 = canonicalize_axis(self.axis1, ndim)
+        axis2 = canonicalize_axis(self.axis2, ndim)
+        if axis1 == axis2:
+            raise ValueError(
+                f"axis1 and axis2 cannot be the same. Received: axis1={axis1}, "
+                f"axis2={axis2}"
```
Comment (hertschuh) [inline keras/src/ops/numpy.py]: Nitpick: display the original `self.axis1` and `self.axis2` in the error, it can be confusing to see something you didn't pass displayed.

**Review by hertschuh — CHANGES_REQUESTED** [review]
Partial review.

**Comment by keerthanakadiri** [comment]
Hi @ChiragSW, Can you please resolve the conflicts? Thank you!

**Comment by ChiragSW** [comment]
## Update

I restarted the implementation from a clean baseline to avoid carrying over the earlier regressions.

Changes made:
- Added a minimal `canonicalize_axes()` helper that delegates to `canonicalize_axis()`.
- Removed duplicated wrapper-level validation and the `get_static_tensor_ndim()` approach.
- Kept public op wrappers simple, with symbolic calls happening directly.
- Moved symbolic axis validation into `compute_output_spec()`.
- Added focused `KerasTensor` tests for invalid symbolic axes.
- Restored unrelated test coverage that had been removed or rewritten earlier.

File: keras/src/ops/nn.py [hunk 15]
```diff
@@ -983,6 +985,11 @@ def call(self, x):
         return backend.nn.log_softmax(x, axis=self.axis)
 
     def compute_output_spec(self, x):
+        if self.axis is not None:
+            if isinstance(self.axis, (tuple, list)):
+                canonicalize_axes(self.axis, len(x.shape))
+            else:
+                canonicalize_axis(self.axis, len(x.shape))
```
Comment (hertschuh) [inline keras/src/ops/nn.py]: You don't need this `if`, `canonicalize_axes` already support the case when it's an int.

**Review by hertschuh — CHANGES_REQUESTED** [review]
One last nitpick:

**Comment by ChiragSW** [comment]
The code format test is failing with an incorrect reasoning.
1. I have not touched image.py file in my PR, still it shows format error in the file.
2. It says error is on line 2167 but image.py is only 1927 lines long
3. Ruff format passes in local system but failing here
Could be a CI error.
@hertschuh the nitpick change was done. Please help with this error.

**Comment by hertschuh** [comment]
> The code format test is failing with an incorrect reasoning.
> 
> 1. I have not touched image.py file in my PR, still it shows format error in the file.
> 2. It says error is on line 2167 but image.py is only 1927 lines long
> 3. Ruff format passes in local system but failing here
>    Could be a CI error.
>    @hertschuh the nitpick change was done. Please help with this error.

Sorry, there was a breakage in the master branch, please rebase.

File: keras/src/ops/numpy_test.py [hunk 16]
```diff
@@ -5125,6 +5140,8 @@ def test_blackman(self):
         x = np.random.randint(1, 100 + 1)
         self.assertAllClose(knp.blackman(x), np.blackman(x))
 
+        self.assertAllClose(knp.Blackman()(x), np.blackman(x))
+
```
Comment (hertschuh) [inline keras/src/ops/numpy_test.py]: Remove, side effect of a bad merge

File: keras/src/ops/numpy_test.py [hunk 17]
```diff
@@ -5133,6 +5150,10 @@ def test_blackman_length_1(self):
         self.assertEqual(out.shape[0], x)
         self.assertAllClose(out, expected)
 
+        out = knp.Blackman()(x_tensor)
+        self.assertEqual(out.shape[0], x)
+        self.assertAllClose(out, expected)
+
```
Comment (hertschuh) [inline keras/src/ops/numpy_test.py]: Remove, side effect of a bad merge

File: keras/src/ops/numpy_test.py [hunk 18]
```diff
@@ -10602,32 +10623,11 @@ def test_nanmin(self, dtype):
         if backend.backend() == "torch" and expected_dtype == "uint32":
             expected_dtype = "int32"
 
-        self.assertEqual(standardize_dtype(knp.nanmin(x).dtype), expected_dtype)
         self.assertEqual(
```
Comment (hertschuh) [inline keras/src/ops/numpy_test.py]: Re-add, side effect of a bad merge

File: keras/src/ops/numpy_test.py [hunk 19]
```diff
@@ -10602,32 +10623,11 @@ def test_nanmin(self, dtype):
...
-
-        self.assertEqual(
-            standardize_dtype(knp.nanpercentile(x, 50).dtype),
-            expected_dtype,
-        )
-        self.assertEqual(
-            standardize_dtype(knp.Nanpercentile().symbolic_call(x, 50).dtype),
-            expected_dtype,
-        )
-
```
Comment (hertschuh) [inline keras/src/ops/numpy_test.py:10630]: Re-add, side effect of a bad merge

**Review by hertschuh — CHANGES_REQUESTED** [review]
It looks like the merge went wrong:

File: keras/src/ops/numpy_test.py [hunk 20]
```diff
@@ -1499,6 +1499,11 @@ def test_blackman(self):
         x = np.random.randint(1, 100 + 1)
         self.assertEqual(knp.blackman(x).shape[0], x)
 
+    def test_blackman_length_1_symbolic_shape(self):
+        x = KerasTensor((1,), dtype="int32")
+        y = knp.blackman(x)
+        self.assertEqual(y.shape, (1,))
+
```
Comment (hertschuh) [inline keras/src/ops/numpy_test.py]: Remove this

File: keras/src/ops/numpy_test.py [hunk 21]
```diff
@@ -10628,6 +10626,24 @@ def test_nanpercentile(self, dtype):
...
+        if backend.backend() == "torch" and expected_dtype == "uint32":
+            expected_dtype = "int32"
+
+        self.assertEqual(standardize_dtype(knp.nanmin(x).dtype), expected_dtype)
+
+        self.assertEqual(
+            standardize_dtype(knp.Nanmin().symbolic_call(x).dtype),
+            expected_dtype,
+        )
+
```
Comment (hertschuh) [inline keras/src/ops/numpy_test.py]: Move back line 10594 like it was before.

**Review by hertschuh — CHANGES_REQUESTED** [review]
There are still some changes from the bad merge:

**Comment by ChiragSW** [comment]
Done!

**Review by hertschuh — APPROVED** [review]
Thanks!
